src/api/utils.js

```diff
@@ -118,40 +168,32 @@ export function useWebSocket(resourceType) {
         return;
       }
       const { operation, payload } = JSON.parse(event.data);
-      queryClient.setQueriesData(resourceType, data => {
-        switch (operation) {
-          case 'Created':
-            const existingResource = data.find(
-              resource => resource.metadata.uid === payload.metadata.uid
-            );
-            if (existingResource) {
-              return data; // payload is stale, ignore it
-            }
-            return [...data, payload];
-          case 'Deleted':
-            return data.filter(
-              resource => resource.metadata.uid !== payload.metadata.uid
-            );
-          case 'Updated':
-            return data.map(resource => {
-              return resource.metadata.uid === payload.metadata.uid &&
-                // only apply the update if it's newer than the version we already have
-                getResourceVersion(payload) > getResourceVersion(resource)
-                ? payload
-                : resource;
-            });
-          default:
-            return data;
-        }
-      });
+      switch (operation) {
+        case 'Created':
+          handleCreated({ payload, queryClient });
+          break;
+        case 'Deleted':
+          handleDeleted({ payload, queryClient });
+          break;
+        case 'Updated':
+          handleUpdated({ payload, queryClient });
+          break;
+        default:
+      }
     };
     webSocket.addEventListener('message', eventListener);
     return () => webSocket.removeEventListener('message', eventListener);
   }, [queryClient, webSocket]);
 }
 
-export function useNamespacedCollection(resourceType, api, params) {
-  const query = useQuery([resourceType, params], () => api(params));
-  useWebSocket(resourceType);
+export function useCollection(kind, api, params) {
+  const query = useQuery([kind, params], () => api(params));
+  useWebSocket({ kind });
+  return query;
+}
+
+export function useResource(kind, api, params) {
```

Comment:
useCollection and useResource seem to be doing the same thing, could we use one?

Reply:
yes we could use one now, but in the near future `useCollection` will be updated to handle automatic chunked loading of large lists (using something like https://github.com/AlanGreene/dashboard/commit/b16612948c62b1ec5d9cb681e14898d7a7a87473#diff-6d1caeaa7a4f194561101db7781e6ff562dd9a95c9cb07852bc69ce6443af15eR165) so splitting them now seemed to make sense since we'll only have to update it here and any container calling `useCollection` will automatically benefit while any using `useResource` will be unaffected